internal/pkg/client/gitlab/fetch.go

```diff
@@ -0,0 +1,198 @@
+// -----------------------------------------------------------------------------
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// -----------------------------------------------------------------------------
+
+package gitlab
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	gitlab "gitlab.com/gitlab-org/api/client-go"
+
+	bomctloptions "github.com/bomctl/bomctl/internal/pkg/options"
+)
+
+type (
+	ProjectProvider interface {
+		GetProject(
+			any,
+			*gitlab.GetProjectOptions,
+			...gitlab.RequestOptionFunc,
+		) (*gitlab.Project, *gitlab.Response, error)
+	}
+
+	BranchProvider interface {
+		GetBranch(
+			any,
+			string,
+			...gitlab.RequestOptionFunc,
+		) (*gitlab.Branch, *gitlab.Response, error)
+	}
+
+	CommitProvider interface {
+		GetCommit(
+			any,
+			string,
+			*gitlab.GetCommitOptions,
+			...gitlab.RequestOptionFunc,
+		) (*gitlab.Commit, *gitlab.Response, error)
+	}
+
+	DependencyListExporter interface {
+		CreateDependencyListExport(
+			int,
+			*gitlab.CreateDependencyListExportOptions,
+			...gitlab.RequestOptionFunc,
+		) (*gitlab.DependencyListExport, *gitlab.Response, error)
+		GetDependencyListExport(
+			int,
+			...gitlab.RequestOptionFunc,
+		) (*gitlab.DependencyListExport, *gitlab.Response, error)
+		DownloadDependencyListExport(int, ...gitlab.RequestOptionFunc) (io.Reader, *gitlab.Response, error)
+	}
+)
+
+var (
+	errInvalidGitLabURL = errors.New("invalid URL for GitLab fetching")
+	errFailedWebRequest = errors.New("web request failed")
+	errForbiddenAccess  = errors.New("the supplied token is missing the read_dependency permission")
+)
+
+func validateHTTPStatusCode(statusCode int) error {
+	if statusCode < http.StatusOK || http.StatusMultipleChoices <= statusCode {
+		return fmt.Errorf("%w. HTTP status code: %d", errFailedWebRequest, statusCode)
+	}
+
+	return nil
+}
+
+func (client *Client) createExport(projectName, branchName string) error {
+	project, response, err := client.GetProject(projectName, nil)
+	if err != nil {
+		return fmt.Errorf("failed to get project info: %w", err)
+	}
+
+	if err := validateHTTPStatusCode(response.StatusCode); err != nil {
+		return err
+	}
+
+	branch, response, err := client.GetBranch(project.ID, branchName)
+	if err != nil {
+		return fmt.Errorf("failed to get branch: %w", err)
+	}
+
+	if err := validateHTTPStatusCode(response.StatusCode); err != nil {
+		return err
+	}
+
+	commit, response, err := client.GetCommit(project.ID, branch.Commit.ID, nil)
```

Comment:
Should we be able to handle a user providing a commit versus always a branch?

Might be a future improvement, but was just thinking about it

Reply:
Hmm. Could be a good idea, although right now I cannot think of a use case where a user might want the SBOM for a commit that is not the latest one in a branch. Good use case for discussion.

Reply:
Also need to make sure the API supports that. IIRC it only supports `ref` and `tag` in the request URL params. `ref` might accept a commit SHA, but I don't know off the top of my head. @lmphil does this PR support specifying a tag?

Reply:
I don't think we should expect the user to provide a pipeline ID, the goal of this client is to get that pipeline ID for them automatically based on a git ref

Reply:
To be resolved in #246